src/command/backup/file.c

```diff
@@ -285,62 +285,84 @@ backupFile(
-                        // Setup the repo file for write. There is no need to write the file atomically (e.g. via a temp file on
-                        // Posix) because checksums are tested on resume after a failed backup. The path does not need to be synced
-                        // for each file because all paths are synced at the end of the backup. It needs to be created in the prior
-                        // context because it will live longer than a single loop when more than one file is being written.
-                        if (write == NULL)
+                        Buffer *const buffer = bufNew(ioBufferSize());
+
+                        do
                         {
-                            MEM_CONTEXT_PRIOR_BEGIN()
+                            // open destination and copy the file only if read something or not bundling
+                            if (ioReadSize(storageReadIo(read), buffer) > 0 || bundleId == 0)
                             {
-                                write = storageNewWriteP(
-                                    storageRepoWrite(), repoFile, .compressible = compressible, .noAtomic = true,
-                                    .noSyncPath = true);
-                                ioWriteOpen(storageWriteIo(write));
+                                // Setup the repo file for write. There is no need to write the file atomically (e.g.
+                                // via a temp file on Posix) because checksums are tested on resume after a failed
+                                // backup. The path does not need to be synced for each file because all paths are
+                                // synced at the end of the backup. It needs to be created in the prior context because
+                                // it will live longer than a single loop when more than one file is being written.
+                                if (write == NULL)
+                                {
+                                    MEM_CONTEXT_PRIOR_BEGIN()
+                                    {
+                                        write = storageNewWriteP(
+                                            storageRepoWrite(), repoFile, .compressible = compressible, .noAtomic = true,
+                                            .noSyncPath = true);
+                                        ioWriteOpen(storageWriteIo(write));
+                                    }
+                                    MEM_CONTEXT_PRIOR_END();
+                                }
+
+                                ioWrite(storageWriteIo(write), buffer);
+                                bufUsedZero(buffer);
                             }
-                            MEM_CONTEXT_PRIOR_END();
                         }
-
-                        // Copy data from source to destination
-                        ioCopyP(storageReadIo(read), storageWriteIo(write));
+                        while (!ioReadEof(storageReadIo(read)));
```

Comment:
why are we reading the file to the end (or why is there a loop?) if it is assumed that::

https://github.com/arenadata/pgbackrest/blob/4fed3fafc06942c14012509cd660a479956cc730/src/command/backup/file.c#L89

Reply:
Original code does not add limit param to ioCopy.

Reply:
Limit is set in read not in ioCopy!
Let, for example, pgFileSize = 100, buffer size = 10 and real file size 1000. If you limit ioCopy it is not work!

Reply:
So if I understood correctly. the original code did not check the file size and copied it full (the actual size at the time of copying)? The new code copies the size that is written in the manifest (if the file size has not been reduced) and we need a loop since we copy through a fixed-size buffer? But why do we check for EOF if it will only be at the end of the file?

Reply:
> the original code did not check the file size and copied it full (the actual size at the time of copying)?

No! It copied with limit in read-object.


> So if I understood correctly. the original code did not check the file size and copied it full (the actual size at the time of copying)? The new code copies the size that is written in the manifest (if the file size has not been reduced) and we need a loop since we copy through a fixed-size buffer? But why do we check for EOF if it will only be at the end of the file?

No! New code is just copy ioCopy internals

Reply:
got it